File: backend/src/main/java/com/blog01/service/PostService.java
package com.blog01.service;

import com.blog01.dto.request.PostRequest;
import com.blog01.dto.response.CommentResponse;
import com.blog01.dto.response.PageResponse;
import com.blog01.dto.response.PostResponse;
import com.blog01.entity.*;
import com.blog01.exception.ForbiddenException;
import com.blog01.exception.ResourceNotFoundException;
import com.blog01.mapper.EntityMapper;
import com.blog01.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collections;
import java.util.List;

@Service
@RequiredArgsConstructor
public class PostService {

    private final PostRepository postRepository;
    private final UserRepository userRepository;
    private final CommentRepository commentRepository;
    private final LikeRepository likeRepository;
    private final ReportRepository reportRepository;
    private final NotificationRepository notificationRepository;
    private final SubscriptionRepository subscriptionRepository;
    private final NotificationService notificationService;
    private final EntityMapper mapper;
    private final SimpMessagingTemplate messagingTemplate;

    public PageResponse<PostResponse> getFeed(User currentUser, int page, int size) {
        List<Long> followingIds = subscriptionRepository.findFollowingIds(currentUser.getId());
        followingIds.add(currentUser.getId());

        if (followingIds.isEmpty()) {
            return emptyPage(page, size);
        }

        Page<Post> posts = postRepository.findFeedPosts(followingIds, PageRequest.of(page, size));
        return toPageResponse(posts, currentUser, false);
    }

    public PageResponse<PostResponse> getUserPosts(String username, User currentUser, int page, int size) {
        User author = userRepository.findByUsername(username)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));
        Page<Post> posts = postRepository.findByAuthorOrderByCreatedAtDesc(author, PageRequest.of(page, size));
        return toPageResponse(posts, currentUser, true);
    }

    public PostResponse getPost(Long id, User currentUser) {
        Post post = findPost(id);
        return toPostResponse(post, currentUser, true);
    }

    @Transactional
    public PostResponse createPost(User currentUser, PostRequest request) {
        Post post = Post.builder()
                .author(currentUser)
                .description(request.getDescription())
                .mediaUrl(request.getMediaUrl())
                .mediaType(request.getMediaType() != null ? request.getMediaType() : MediaType.NONE)
                .build();
        post = postRepository.save(post);

        notificationService.notifyFollowersOfNewPost(currentUser, post);

        PostResponse response = toPostResponse(post, currentUser, true);
        messagingTemplate.convertAndSend("/topic/feed", response);
        return response;
    }

    @Transactional
    public PostResponse updatePost(Long id, User currentUser, PostRequest request) {
        Post post = findPost(id);
        assertOwner(post, currentUser);
        post.setDescription(request.getDescription());
        if (request.getMediaUrl() != null) post.setMediaUrl(request.getMediaUrl());
        if (request.getMediaType() != null) post.setMediaType(request.getMediaType());
        post = postRepository.save(post);
        return toPostResponse(post, currentUser, true);
    }

    @Transactional
    public void deletePost(Long id, User currentUser) {
        Post post = findPost(id);
        if (!post.getAuthor().getId().equals(currentUser.getId()) && currentUser.getRole() != Role.ADMIN) {
            throw new ForbiddenException("Not allowed to delete this post");
        }
        commentRepository.deleteByPost(post);
        likeRepository.deleteByPost(post);
        reportRepository.deleteByReportedPost(post);
        notificationRepository.deleteByRelatedPostId(post.getId());
        postRepository.delete(post);
    }

    private Post findPost(Long id) {
        return postRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Post not found"));
    }

    private void assertOwner(Post post, User user) {
        if (!post.getAuthor().getId().equals(user.getId())) {
            throw new ForbiddenException("Not allowed to modify this post");
        }
    }

    private PostResponse toPostResponse(Post post, User currentUser, boolean includeComments) {
        long likes = likeRepository.countByPost(post);
        long comments = commentRepository.countByPost(post);
        boolean liked = likeRepository.existsByPostAndUser(post, currentUser);
        List<CommentResponse> commentList = includeComments
                ? commentRepository.findByPostOrderByCreatedAtAsc(post).stream()
                    .map(mapper::toCommentResponse).toList()
                : List.of();
        return mapper.toPostResponse(post, currentUser, likes, comments, liked, commentList);
    }

    private PageResponse<PostResponse> toPageResponse(Page<Post> page, User currentUser, boolean includeComments) {
        List<PostResponse> content = page.getContent().stream()
                .map(p -> toPostResponse(p, currentUser, includeComments))
                .toList();
        return PageResponse.<PostResponse>builder()
                .content(content)
                .page(page.getNumber())
                .size(page.getSize())
                .totalElements(page.getTotalElements())
                .totalPages(page.getTotalPages())
                .last(page.isLast())
                .build();
    }

    private PageResponse<PostResponse> emptyPage(int page, int size) {
        return PageResponse.<PostResponse>builder()
                .content(Collections.emptyList())
                .page(page)
                .size(size)
                .totalElements(0)
                .totalPages(0)
                .last(true)
                .build();
    }
}

File: frontend/src/app/core/services/post.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { tap } from 'rxjs/operators';
import { environment } from '../../../environments/environment';
import { PageResponse, Post } from '../models';
import { AuthService } from './auth.service';
import { UserStatsService } from './user-stats.service';

@Injectable({ providedIn: 'root' })
export class PostService {
  constructor(
    private http: HttpClient,
    private auth: AuthService,
    private userStats: UserStatsService
  ) {}

  getFeed(page = 0, size = 10) {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<PageResponse<Post>>(`${environment.apiUrl}/posts/feed`, { params });
  }

  getUserPosts(username: string, page = 0, size = 10) {
    const params = new HttpParams().set('page', page).set('size', size);
    return this.http.get<PageResponse<Post>>(`${environment.apiUrl}/posts/user/${username}`, { params });
  }

  getPost(id: number) {
    return this.http.get<Post>(`${environment.apiUrl}/posts/${id}`);
  }

  createPost(data: { description: string; mediaUrl?: string; mediaType?: string }) {
    return this.http.post<Post>(`${environment.apiUrl}/posts`, data).pipe(
      tap(() => {
        const user = this.auth.currentUser();
        if (user) {
          this.userStats.adjustPostCount(user.username, 1);
        }
      })
    );
  }

  updatePost(id: number, data: { description: string; mediaUrl?: string | null; mediaType?: string }) {
    return this.http.put<Post>(`${environment.apiUrl}/posts/${id}`, data);
  }

  deletePost(id: number, authorUsername?: string) {
    return this.http.delete<void>(`${environment.apiUrl}/posts/${id}`).pipe(
      tap(() => {
        const username = authorUsername ?? this.auth.currentUser()?.username;
        if (username) {
          this.userStats.adjustPostCount(username, -1);
        }
      })
    );
  }

  toggleLike(postId: number) {
    return this.http.post<{ liked: boolean; likesCount: number }>(
      `${environment.apiUrl}/posts/${postId}/likes`, {}
    );
  }

  getComments(postId: number) {
    return this.http.get<import('../models').Comment[]>(`${environment.apiUrl}/posts/${postId}/comments`);
  }

  addComment(postId: number, content: string) {
    return this.http.post<import('../models').Comment>(
      `${environment.apiUrl}/posts/${postId}/comments`, { content }
    );
  }

  deleteComment(postId: number, commentId: number) {
    return this.http.delete<void>(`${environment.apiUrl}/posts/${postId}/comments/${commentId}`);
  }
}
